src/modules/sdexec/sdexec.c

```diff
@@ -0,0 +1,790 @@
+        json_decref (ctx->local_uri);
+        free (ctx->unit_prefix);
+        free (ctx);
+        errno = saved_errno;
+    }
+}
+
+static struct sdexec_ctx *sdexec_ctx_create (flux_t *h)
+{
+    struct sdexec_ctx *ctx;
+    const char *s;
+
+    if (!(ctx = calloc (1, sizeof (*ctx))))
+        return NULL;
+    ctx->h = h;
+    if (flux_get_rank (h, &ctx->rank) < 0)
+        goto error;
+    if (!(s = flux_attr_get (h, "local-uri")))
+        goto error;
+    if (!(ctx->local_uri = json_string (s))) {
+        errno = ENOMEM;
+        goto error;
+    }
+    if (!(ctx->requests = flux_msglist_create ())
+        || !(ctx->kills = flux_msglist_create ()))
+        goto error;
+    return ctx;
+error:
+    sdexec_ctx_destroy (ctx);
+    return NULL;
+}
+
+static void delete_message (struct flux_msglist *msglist,
+                            const flux_msg_t *msg)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        if (msg == m) {
+            flux_msglist_delete (msglist);
+            return;
+        }
+        m = flux_msglist_next (msglist);
+    }
+}
+
+static const flux_msg_t *lookup_message_bypath (struct flux_msglist *msglist,
+                                                const char *path)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        struct sdproc *proc = flux_msg_aux_get (m, "sdproc");
+        if (proc && streq (proc->path, path))
+            return m;
+        m = flux_msglist_next (msglist);
+    }
+    return NULL;
+}
+
+static const flux_msg_t *lookup_message_bypid (struct flux_msglist *msglist,
+                                               pid_t pid)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        struct sdproc *proc = flux_msg_aux_get (m, "sdproc");
+        if (sdexec_unit_pid (proc->unit) == pid)
+            return m;
+        m = flux_msglist_next (msglist);
+    }
+    return NULL;
+}
+
+static const flux_msg_t *lookup_message_byaux (struct flux_msglist *msglist,
+                                               const char *name,
+                                               void *value)
+{
+    const flux_msg_t *m;
+
+    m = flux_msglist_first (msglist);
+    while (m) {
+        if (flux_msg_aux_get (m, name) == value)
+            return m;
+        m = flux_msglist_next (msglist);
+    }
+    return NULL;
+}
+
+/* Send the streaming response IFF unit cleanup is complete and EOFs have
+ * been sent.  Channel EOF and cleanup might(?) complete out of order so
+ * call this from unit and channel callbacks.
+ */
+static void finalize_exec_request (struct sdexec_ctx *ctx,
+                                   struct sdproc *proc,
+                                   struct unit *unit)
+{
+    if (sdexec_unit_state (unit) == STATE_INACTIVE
+        && sdexec_unit_substate (unit) == SUBSTATE_DEAD
+        && proc->out_eof_sent
+        && proc->err_eof_sent) {
```

Comment:
should be `!proc->out || proc->out_eof_sent` (and stderr too) b/c STDOUT/STDERR channels can be optionally set?